Run all C/C++ code under `src` through `clang-format` and add ongoing enforcement

jeamland opened the pull request:
This could be considered a very mild troll of @cgwalters.

This change would allow anyone using an editor like Visual Studio Code (or many others) to enable formatting on save using clang-format's definition of GNU style.

I'm more than happy to tweak the style configuration if needed/desired.

cgwalters commented:
Heh.  From my PoV the main downside is the hit to `git annotate` across the board - anytime in the future we're doing that we'll almost certainly need to go back to the previous commit.  The other downside is that all outstanding PRs that touch C/C++ will conflict with this (but there aren't many).  Are there any other downsides I'm not thinking of?  I don't honestly have much experience with `clang-format`.  I will try to play with this PR locally for a bit.

The current code has its inconsistencies, but it's *mostly* consistent.

OTOH, the value of having editor auto-format working is not to be underestimated.

The alternative path I think is that we basically just keep trying to slowly drain our C/C++ into Rust.

I'm curious what other contributors think.

travier commented:
We could also add an ignoreRevsFile with the commit doing the mass styling: https://git-scm.com/docs/git-blame#Documentation/git-blame.txt---ignore-revs-fileltfilegt

travier commented:
But the biggest problem with this PR is that it's really hard to review.

lucab commented:
/ok-to-test

lucab commented:
I don't have style preferences nor much experience with `clang-format`, but as long as it works reliably and can be checked in CI I'm very happy to have this!

jeamland commented:
> I'm personally OK with this, and don't mind too much the `git annotate` hit. It shouldn't affect `git log -S` at least, which I find is often more useful. Awesome that you've added it to the codestyle tests! 👍
> 
> Two tweaks I'd suggest:
> 
> * Can we increase the line length limit? It looks like it's currently 80 which is really tiny. Let's say... at least 100? Cool with higher too if folks prefer.
> * Can we make function argument declarations stay on separate lines? E.g. keep
> 
> ```c
> gboolean rpmostree_importer_run (RpmOstreeImporter *self,
>                                  char             **out_csum,
>                                  GCancellable      *cancellable,
>                                  GError           **error)
> ```
> 
> instead of
> 
> ```c
> rpmostree_importer_run (RpmOstreeImporter *self, char **out_csum,
>                         GCancellable *cancellable, GError **error)
> ```
> 
> Looking at the changes, I suspect those two tweaks should cut the diff by a good factor too.

So I had a bit of a look at the various frobs `clang-format` provides and there's no real way to achieve exactly what you're after here.

Extending the line length limit is easy. The default is 79 columns, setting it to 99 is easy, and setting it higher is fine if so desired.

When it comes to the parameter formatting it becomes a bit trickier.

So starting with this, which is what is currently in `main`:

```c
gboolean
rpmostree_importer_run (RpmOstreeImporter *self,
                        char             **out_csum,
                        GCancellable      *cancellable,
                        GError           **error)
```

This is strict GNU style (per `clang-format`'s interpretation):

```c
rpmostree_importer_run (RpmOstreeImporter *self, char **out_csum,
                        GCancellable *cancellable, GError **error)
```

Both of those are what's already shown above. We can then shift to a custom configuration that looks like this:

```yaml
---
Language:          Cpp
BasedOnStyle:      GNU
BinPackParameters: false
ColumnLimit:       99
```

Which gets us this:

```c
gboolean
rpmostree_importer_run (RpmOstreeImporter *self,
                        char **out_csum,
                        GCancellable *cancellable,
                        GError **error)
```

That's closer, but still not exactly the same. The closest I think I can get is this, where we add this to the configuration:

```yaml
AlignConsecutiveDeclarations: AcrossEmptyLinesAndComments
```

And get this:

```c
gboolean
rpmostree_importer_run (RpmOstreeImporter *self,
                        char **            out_csum,
                        GCancellable *     cancellable,
                        GError **          error)
```

The issue with that last bit is that it will also change things like this (currently what's in `main`):

```c
struct RpmOstreeImporter
{
  GObject parent_instance;
  OstreeRepo *repo;
  OstreeSePolicy *sepolicy;
  struct archive *archive;
  int fd;
  Header hdr;
  rpmfi fi;
  off_t cpio_offset;
  // Hashmap of file-overrides from RPM header:
  //  - [K] absolute full path of the file
  //  - [V] iterator index in RPM header for this file
  GHashTable *rpmfi_overrides;
  // Hashset of all file entries marked as 'doc' in an RPM;
  // each key is the absolute full path of the file.
  GHashTable *doc_files;
  // Hashset of filepath entries which are direct children of /opt;
  // each key is a plain path fragment, e.g. 'foo' for '/opt/foo/bar'.
  GHashTable *opt_direntries;
  // Hashset of directories which got moved from '/var/lib' to '/usr/lib';
  // each key is a plain directory name, e.g. 'foo' for '/var/lib/foo/'.
  GHashTable *varlib_direntries;
  GString *tmpfiles_d;
  RpmOstreeImporterFlags flags;
  DnfPackage *pkg;
  char *hdr_sha256;

  char *ostree_branch;
};
```

into this:

```c
struct RpmOstreeImporter
{
  GObject                parent_instance;
  OstreeRepo *           repo;
  OstreeSePolicy *       sepolicy;
  struct archive *       archive;
  int                    fd;
  Header                 hdr;
  rpmfi                  fi;
  off_t                  cpio_offset;
  // Hashmap of file-overrides from RPM header:
  //  - [K] absolute full path of the file
  //  - [V] iterator index in RPM header for this file
  GHashTable *           rpmfi_overrides;
  // Hashset of all file entries marked as 'doc' in an RPM;
  // each key is the absolute full path of the file.
  GHashTable *           doc_files;
  // Hashset of filepath entries which are direct children of /opt;
  // each key is a plain path fragment, e.g. 'foo' for '/opt/foo/bar'.
  GHashTable *           opt_direntries;
  // Hashset of directories which got moved from '/var/lib' to '/usr/lib';
  // each key is a plain directory name, e.g. 'foo' for '/var/lib/foo/'.
  GHashTable *           varlib_direntries;
  GString *              tmpfiles_d;
  RpmOstreeImporterFlags flags;
  DnfPackage *           pkg;
  char *                 hdr_sha256;

  char *                 ostree_branch;
};
```

So you don't end up saving too much diff unfortunately. Personally I like aligning structure field names like that but that's my aesthetic preference. Up to you whether you want to go in on that or not.

jeamland commented:
I've re-pushed this with the following configuration:

```yaml
---
Language: Cpp
BasedOnStyle: GNU
AlignConsecutiveDeclarations: AcrossEmptyLinesAndComments
BinPackParameters: false
ColumnLimit: 99
```

Let me know what you think.

cgwalters commented:
OK so I definitely agree on at least 99 char column limit.

Beyond that just to repeat I would rather spend my mental energy on conversion to Rust, so I don't have a really strong opinion on all the details here.  Reading over it, there's things that look slightly odd but it overall looks OK.

I think aligning with an existing standard is good; I originally chose GNU just based on my background (one of the first FOSS projects I contributed to was Emacs), and I spent a lot of time in GNOME which also has a lot of GNU-formatted C, most notably glib/gtk+ (but not exclusively, other GNOME projects are Linux style).

I think the "aligned parameter list" thing actually came from GNOME?  Glancing at e.g. GCC, they don't seem to do that.  Another GNU-formatted project I recently looked at is [crun](https://github.com/containers/crun/) and they also don't align parameter lists.  Oh wow but their [clang-format file is enormous](https://github.com/containers/crun/blob/main/.clang-format) ?  @giuseppe what's going on with that?  Did you have a specific rationale for all those knobs?

I think I'd lean towards just:

```
Language: Cpp
BasedOnStyle: GNU
ColumnLimit: 99
```

i.e. keep it simple.

cgwalters commented:
Although to be clear, I'm also fine with the proposed configuration.  Maybe we should vote?

Let's call [this configuration](https://github.com/coreos/rpm-ostree/pull/3475#issuecomment-1055003827) Option 1, [this one](https://github.com/coreos/rpm-ostree/pull/3475#issuecomment-1055524999) as Option 2 - are there any other tweaks?  Maybe react to the comments with a :+1: for which is preferred?  I am actually pretty neutral.

cgwalters commented:
Why is CI failing btw?  It seems like there's some clang format option you used that controls where e.g. `*` goes that isn't set in the config?

giuseppe commented:
> @giuseppe what's going on with that? Did you have a specific rationale for all those knobs?

not really.  I've used `clang-format --dump-config --style=gnu` as a starting point and did some tweaks to minimize changes to the existing code

cgwalters commented:
@jeamland I believe there's consensus here to do this, so (not urgent) but feel free to rebase this and we can get it in.

jeamland commented:
> @jeamland I believe there's consensus here to do this, so (not urgent) but feel free to rebase this and we can get it in.

No worries. Did you want me to flip the config back to straight GNU with a 100-column line limit?

cgwalters commented:
> No worries. Did you want me to flip the config back to straight GNU with a 100-column line limit?

Yeah; as best I can tell, there was indeed weak consensus for that.

jeamland commented:
Ok, I've adjusted the clang-format config to be GNU with a 100-column limit, rebased, and ensured that the result builds. I've noticed that some of the CI runs seem to decide something needs a rebase anyway so I'm not sure what the right way forward is if they fail again.

cgwalters commented:
OK, the last commit is obviously impossible to review manually, but since it's nicely split up it's easy to verify, which is what I did locally:

```
$ <git remote add jeamland, fetch branch, check out locally>
$ git checkout -b my-clang-format
$ git reset --hard HEAD^ # Undo the format commit
$ git ls-files 'src/**.c' 'src/**.cxx' 'src/**.h' 'src/**.hpp' | xargs clang-format -i
$ git commit -a -m format
$ git diff jeamland-clang-format..
$
```

Thank you for this patch!

cgwalters commented:
Followup in https://github.com/coreos/rpm-ostree/pull/3498
